### PR TITLE
Enable communication to management API over HTTPS. Fixes #22953

### LINKED ISSUE #22953: rabbitmq_queue has no option to use HTTPS when accessing RabbitMQ management API
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
rabbitmq_queue

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 6cea89299a) last updated 2017/03/24 14:57:26 (GMT -400)
  config file = /home/redacted/.ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### CONFIGURATION
< destination_variable = private_dns_name
< vpc_destination_variable = private_ip_address

##### OS / ENVIRONMENT
Controller:
```
> cat /etc/lsb-release 
DISTRIB_ID=LinuxMint
DISTRIB_RELEASE=18
DISTRIB_CODENAME=sarah
DISTRIB_DESCRIPTION="Linux Mint 18 Sarah"
```
Target:
```
> cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.5 LTS"
```

##### SUMMARY
The module hardcodes "http" in the URI which it generates which prevents it from accessing systems which are expecting HTTPS
`url = "http://%s:%s/api/queues/%s/%s" % (`

##### STEPS TO REPRODUCE
Use the module to create a queue on a RabbitMQ instance which requires HTTPS for the management interface.

```yaml
- name: Create queues
  rabbitmq_queue:
    name: testq
    vhost: "{{ rmq_vhost }}"
    login_user: "{{ rmq_user }}"
    login_password: "{{ rmq_pass }}"
    login_host: "{{ ansible_fqdn }}"
    login_port: 15671
    durable: True
```
##### EXPECTED RESULTS
The queue 'testq' should be created under vhost "{{ rmq_vhost }}"'  on the RabbitMQ instance running at  "{{ ansible_fqdn }}" who's management interface is using HTTPS on port 15671

##### ACTUAL RESULTS

The API call fails because it's using HTTP instead of HTTPS

```
Using module file /home/redacted/Ansible/ansible/lib/ansible/modules/messaging/rabbitmq_queue.py
<172.16.191.87> ESTABLISH SSH CONNECTION FOR USER: automation
<172.16.191.87> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=automation -o ConnectTimeout=10 -o ControlPath=/home/redacted/.ansible/cp/999765f0fe 172.16.191.87 '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-cfjgpsspmnwpsvntxvjablneyhyiyfvr; redacted/pyenvs/ansible/bin/python'"'"'"'"'"'"'"'"' && sleep 0'"'"''
<172.16.191.87> (1, '', 'Traceback (most recent call last):\n  File "/tmp/ansible_qDVLbU/ansible_module_rabbitmq_queue.py", line 277, in <module>\n    main()\n  File "/tmp/ansible_qDVLbU/ansible_module_rabbitmq_queue.py", line 167, in main\n    r = requests.get( url, auth=(module.params[\'login_user\'],module.params[\'login_password\']))\n  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/api.py", line 70, in get\n    return request(\'get\', url, params=params, **kwargs)\n  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/api.py", line 56, in request\n    return session.request(method=method, url=url, **kwargs)\n  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/sessions.py", line 488, in request\n    resp = self.send(prep, **send_kwargs)\n  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/sessions.py", line 609, in send\n    r = adapter.send(request, **kwargs)\n  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/adapters.py", line 487, in send\n    raise ConnectionError(e, request=request)\nrequests.exceptions.ConnectionError: HTTPConnectionPool(host=\'testhost.some.domain\', port=15671): Max retries exceeded with url: /api/queues/%2Fhec%2Ftest/recording.upload.processing (Caused by NewConnectionError(\'<requests.packages.urllib3.connection.HTTPConnection object at 0x7fca0cc12310>: Failed to establish a new connection: [Errno 111] Connection refused\',))\n')
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_qDVLbU/ansible_module_rabbitmq_queue.py", line 277, in <module>
    main()
  File "/tmp/ansible_qDVLbU/ansible_module_rabbitmq_queue.py", line 167, in main
    r = requests.get( url, auth=(module.params['login_user'],module.params['login_password']))
  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/api.py", line 70, in get
    return request('get', url, params=params, **kwargs)
  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/api.py", line 56, in request
    return session.request(method=method, url=url, **kwargs)
  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/sessions.py", line 488, in request
    resp = self.send(prep, **send_kwargs)
  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/sessions.py", line 609, in send
    r = adapter.send(request, **kwargs)
  File "redacted/pyenvs/ansible/local/lib/python2.7/site-packages/requests/adapters.py", line 487, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='testhost.some.domain', port=15671): Max retries exceeded with url: /api/queues/%2Ffubar/testq (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x7fca0cc12310>: Failed to establish a new connection: [Errno 111] Connection refused',))

```

### PR DESCRIPTION
**Pull request by Jorge-Rodriguez**
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
rabbitmq utils

##### ANSIBLE VERSION
```
ansible 2.2.0 (devel 4dddea69c2) last updated 2016/11/09 16:33:00 (GMT +300)
  lib/ansible/modules/core:  not found - use git submodule update --init lib/ansible/modules/core
  lib/ansible/modules/extras:  not found - use git submodule update --init lib/ansible/modules/extras
  config file = 
  configured module search path = Default w/o overrides

```

##### SUMMARY
These files complement PR ansible/ansible-modules-extras#2872

### CODE REVIEW

**Comment by mattclay** [comment]
@Jorge-Rodriguez Please rebase your branch and update this PR so tests can be re-run.

**Comment by mattclay** [comment]
The branch this PR is based on is too old to work with CI. Once it is rebased, tests will re-run automatically.

**Comment by calfonso** [comment]
@Jorge-Rodriguez could you rebase this so we can run the CI against it? If you're no longer interested in having it merged, you can close it. Otherwise, we'll watch for the rebase and review/merge.

**Comment by Jorge-Rodriguez** [comment]
@calfonso sorry for the delay, you caught me at the end of my holidays. This should be up to date and ready for review/merge. I added a companion commit from an old PR to ansible-modules-core

**Comment by Jorge-Rodriguez** [comment]
This is now ready_for_review
@calfonso @bcoca can we look into merging this before it sits gathering dust for another year and needs to be rebased again?

**Comment by calfonso** [comment]
@Jorge-Rodriguez A community supported module requires a community reviewer. PRs without an active reviewer will sit longer. Waiting for a review from @chrishoffman @manuel-sousa @romanek-adam

**Comment by Jorge-Rodriguez** [comment]
This is just great. Another rebase request and now my PR has failed checks because of an error in an unrelated part of the code.
@calfonso @bcoca @chrishoffman @manuel-sousa @romanek-adam can we have this looked at and merged already? There is a demand for this feature.

**Comment by mattclay** [comment]
The CI issue has been fixed. Tests are now passing.

**Comment by Jorge-Rodriguez** [comment]
@chrishoffman @manuel-sousa @romanek-adam No movement at all on this still?

**Comment by zwindler** [comment]
Hi, this is exactly what I need!

Since tests are passing and everybody seem to be OK on this, can this be merged?

Regards

**Comment by Jorge-Rodriguez** [comment]
@mattclay yes, please, this has been sitting here for longer than I can remember.

**Comment by Akasurde** [comment]
@Jorge-Rodriguez Could you please rebase in order to remove conflicts ? Ping me once it is done.

**Comment by Jorge-Rodriguez** [comment]
@Akasurde rebased

File: lib/ansible/modules/messaging/rabbitmq_queue.py
Comment (Akasurde) [inline lib/ansible/modules/messaging/rabbitmq_queue.py]: This is extraneous here.

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by Jorge-Rodriguez** [comment]
@Akasurde check removed

**Review by Akasurde — APPROVED** [review]

**Comment by Akasurde** [comment]
cc @resmo Could you please review this, thanks?

**Comment by jmturwy** [comment]
for the love of god can this be merged?

**Comment by Akasurde** [comment]
@Jorge-Rodriguez Thanks for the contribution and patience. And Sorry for delay.

File: lib/ansible/modules/messaging/rabbitmq_binding.py
Comment (hyperized) [inline lib/ansible/modules/messaging/rabbitmq_binding.py:245]: Bit late to the party, but `self.cacert` is never set, only `self.verify`, so that seems a bug. And as far as I can see verify is not a kwarg that request.post actually uses, it only accepts `client_cert` and `client_key`.

Comment (Akasurde) [reply lib/ansible/modules/messaging/rabbitmq_binding.py:245]: Yep, it seems that it introduced a bug. @Jorge-Rodriguez Can you please check this ?

Comment (Jorge-Rodriguez) [reply lib/ansible/modules/messaging/rabbitmq_binding.py:245]: Interesting, good catch.
Fix on the way

Comment (Jorge-Rodriguez) [reply lib/ansible/modules/messaging/rabbitmq_binding.py:245]: @hyperized `requests.post` does accept `verify`
https://github.com/requests/requests/blob/master/requests/api.py#L111
https://github.com/requests/requests/blob/master/requests/api.py#L41
https://github.com/requests/requests/blob/master/requests/sessions.py#L490

Comment (hyperized) [reply lib/ansible/modules/messaging/rabbitmq_binding.py:245]: @Jorge-Rodriguez Thank you for verifying, I wasn't sure!

Comment (hyperized) [reply lib/ansible/modules/messaging/rabbitmq_binding.py:245]: Regarding `requests`, isn't this module using: https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/urls.py#L888 ?

Comment (Jorge-Rodriguez) [reply lib/ansible/modules/messaging/rabbitmq_binding.py:245]: Well, now I'm not sure. But I've been using this code for two years now and it does work.
Let me see.

**Edit:**
Actually no, it's not using urls.py but the python requests package:
https://github.com/Jorge-Rodriguez/ansible/blob/df9a0815d1bc591b61cb2185d4ae4a7cf6064132/lib/ansible/modules/messaging/rabbitmq_binding.py#L81
https://github.com/Jorge-Rodriguez/ansible/blob/df9a0815d1bc591b61cb2185d4ae4a7cf6064132/lib/ansible/modules/messaging/rabbitmq_binding.py#L130

**Pull request merged**
